pex/resolve/lock_resolver.py

```diff
@@ -42,40 +48,117 @@ def __init__(
         try:
             with open(path, "wb") as fp, self._url_fetcher.get_body_stream(url) as stream:
                 for chunk in iter(lambda: stream.read(self._BUFFER_SIZE), b""):
                     fp.write(chunk)
-                    digest_check.update(chunk)
-                    digest_internal.update(chunk)
+                    digest.update(chunk)
+            return artifact.filename
         except IOError as e:
-            raise ResultError(Error(str(e)))
-
-        actual_hash = digest_check.hexdigest()
-        if artifact.fingerprint.hash != actual_hash:
-            raise ResultError(
-                Error(
-                    "Expected {algorithm} hash of {expected_hash} but hashed to "
-                    "{actual_hash}.".format(
-                        algorithm=artifact.fingerprint.algorithm,
-                        expected_hash=artifact.fingerprint.hash,
-                        actual_hash=actual_hash,
-                    )
+            return Error(str(e))
+
+
+class VCSArtifactDownloadManager(DownloadManager[VCSArtifact]):
+    def __init__(
+        self,
+        indexes=None,  # type: Optional[Sequence[str]]
+        find_links=None,  # type: Optional[Sequence[str]]
+        resolver_version=None,  # type: Optional[ResolverVersion.Value]
+        network_configuration=None,  # type: Optional[NetworkConfiguration]
+        cache=None,  # type: Optional[str]
+        use_pep517=None,  # type: Optional[bool]
+        build_isolation=True,  # type: bool
+        pex_root=None,  # type: Optional[str]
+    ):
+        super(VCSArtifactDownloadManager, self).__init__(pex_root=pex_root)
+        self._indexes = indexes
+        self._find_links = find_links
+        self._resolver_version = resolver_version
+        self._network_configuration = network_configuration
+        self._cache = cache
+        self._use_pep517 = use_pep517
+        self._build_isolation = build_isolation
+
+    def save(
+        self,
+        artifact,  # type: VCSArtifact
+        project_name,  # type: ProjectName
+        dest_dir,  # type: str
+        digest,  # type: HintedDigest
+    ):
+        # type: (...) -> Union[str, Error]
+
+        requirement = artifact.as_unparsed_requirement(project_name)
+        downloaded_vcs = resolver.download(
+            requirements=[requirement],
+            transitive=False,
+            indexes=self._indexes,
+            find_links=self._find_links,
+            resolver_version=self._resolver_version,
+            network_configuration=self._network_configuration,
+            cache=self._cache,
+            use_wheel=False,
+            prefer_older_binary=False,
+            use_pep517=self._use_pep517,
+            build_isolation=self._build_isolation,
+            max_parallel_jobs=1,
+        )
+        if len(downloaded_vcs.local_distributions) != 1:
+            return Error(
+                "Expected 1 artifact for an intransitive download of {requirement}, found "
```

Comment:
Rather than "intransitive", would directly saying "VCS" work? That may be more clear, intransitive seems like an implementation detail fwict

Reply:
This is effectively an assert. It should never happen. If a user sees this they are up a creek and there is absolutely nothing they can do except report a bug. The only useful information is the backtrace, which will lead to this code anyhow.